src/Gaffer/Plug.cpp

```diff
@@ -691,6 +691,41 @@ void Plug::parentChanged( Gaffer::GraphComponent *oldParent )
 	popDirtyPropagationScope();
 }
 
+void Plug::childrenReordered( const std::vector<size_t> &oldIndices )
+{
+	// Reorder the children of our outputs to match our new order. We disable
+	// undo while we do this, because `childrenReordered()` will be called again
+	// when the original action is undone anyway.
+	UndoScope undoDisabler( ancestor<ScriptNode>(), UndoScope::Disabled );
+	for( auto output : m_outputs )
+	{
+		if( output->children().size() != oldIndices.size() )
+		{
+			IECore::msg(
+				IECore::Msg::Warning, "Plug::childrenReordered",
+				boost::format( "Not reordering output \"%1%\" because its size doesn't match the input" ) % output->fullName()
+			);
+			continue;
+		}
+		GraphComponent::ChildContainer children; children.reserve( oldIndices.size() );
+		for( auto i : oldIndices )
+		{
+			children.push_back( output->getChild( i ) );
+		}
```

Comment:
I was contemplating whether it was a good idea to have the inner API deal with integer permutation lists, but the outer API deal with lists of children.  I had pretty well convinced myself that it was the right decision ( the list of children is usually more convenient from outside, the integer permutation list is more useful internally, and using different conventions in the internal API is fine ), but this use case here suggests possibly we should expose both?  It's a bit silly to have to convert from integer list to child list here, just so that reorderChildren can translate back before actually appyling it.

Reply:
I prefer an API where there is one way of doing something, and I think lists of children are clearer and easier to use for the majority of cases :

```
# Can this be expressed as neatly with indices?
p.reorderChildren( reversed( p.children() ) )
p.reorderChildren( sorted( p.children(), key = lambda c : c.getName() ) )
```

If performance were to become an issue that might tip the scales, but note that even when passing indices they will need a validation step anyway, not unlike the current version. I think I'd consider a protected `reorderChildren( indices )` variant before I would consider adding public API. This could be for trusted callers like Plug, meaning we could skip the validation entirely.

Reply:
I was pretty on the fence myself, this reasoning seems fine.